mcl/ndarray.py

```diff
@@ -66,10 +70,110 @@ def __setitem__(self, idx: _Indices, value: T):
 
     def __getitem__(self, idx: _Indices) -> Array[T] | Generic:
         if not isinstance(idx, tuple):
             idx = (idx,)
-        idx = tuple_cast(intp, idx)
-        return self.dtype.type.from_memory(self.data, idx)
+
+        if len(idx) < len(self.shape):
+            idx = idx + (slice(None),) * (len(self.shape) - len(idx))
+
+        if any(isinstance(i, slice) for i in idx):
+            new_shape, new_strides, new_offset = self.new_arrayinfo(idx)
+            new_memref = self.data.view(new_shape, new_strides, new_offset)
+            return Array(dtype=self.dtype, data=new_memref)
+        else:
+            # All indices are integers
+            idx = tuple_cast(intp, idx)
+            # TODO: There's no assertion that checks if idx is within bounds.
+            return self.dtype.type.from_memory(self.data, idx)
+
+    def broadcast_to(self, shape: tuple[intp, ...]) -> None:
+        # This function can also serve as a assertion
+        new_shape = self.broadcast_shapes(self.shape, shape)
+        assert new_shape == shape
+
+        old_shape = self.shape
+        old_strides = self.strides
+        new_ndim = len(new_shape)
+        ndim_diff = len(new_shape) - len(old_shape)
+        extended_shape = (intp(1),) * ndim_diff + old_shape
+        extended_strides = (intp(0),) * ndim_diff + old_strides
+        new_strides = [intp(0)] * new_ndim
+        for i in range(new_ndim):
+            if new_shape[i] == extended_shape[i]:
+                new_strides[i] = extended_strides[i] * (new_shape[i] // extended_shape[i])
+
+        new_strides = tuple(new_strides)
+
+        self.data = self.data.view(new_shape, new_strides, self.data.offset)
+
+    @classmethod
+    def broadcast_shapes(cls, *shapes) -> tuple[intp, ...]:
+        if len(shapes) == 0:
+            raise ValueError("At least one shape is required")
+
+        max_ndim = max(map(len, shapes))
+        identity = intp(1)
+
+        # The length can be determined at compile time
+        # So something like tuple setitem can be used
+        # here instead of building a list.
+        result = [identity] * max_ndim
+
+        for i in range(1, max_ndim + 1):
+            curr_axis_shape = identity
+            for shape in shapes:
+                if len(shape) + 1 > i:
+                    if shape[-i] == identity:
+                        pass
+                    elif curr_axis_shape is identity:
+                        curr_axis_shape = shape[-i]
+                    elif curr_axis_shape != shape[-i]:
+                        raise ValueError("Shapes are not broadcastable")
+            result[-i] = curr_axis_shape
+
+        return tuple(result)
+
+    def new_arrayinfo(self, idx: _Indices) -> tuple[intp, ...]:
+        res_ndim = self.ndim
+
+        for idx_ in idx:
+            if isinstance(idx_, (int, intp)):
+                res_ndim -= intp(1)
+
+        res_shape = [intp(0)] * res_ndim
+        res_strides = [intp(0)] * res_ndim
```

Comment:
These will need to be a VM op for now. Later, there should be a rewrite pass for it.